spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -165,50 +180,65 @@ private boolean isNativeEncodingNotSet(ProducerProperties producerProperties, Co
+					"Multiple  beans of type 'PartitionKeyExtractorStrategy' found. " + this.partitionKeyExtractors + ". Please "
+							+ "use 'spring.cloud.stream.bindings.output.producer.partitionKeyExtractorName' property to specify "
+							+ "the name of the bean to be used.");			
+			partitionKeyExtractor = CollectionUtils.isEmpty(this.partitionKeyExtractors) ? 
+					null : this.partitionKeyExtractors.values().iterator().next();
+		}
+		return partitionKeyExtractor;
 	}
 
-	private PartitionSelectorStrategy getPartitionSelectorStrategy(
-			ProducerProperties producerProperties) {
+	@SuppressWarnings("deprecation")
+	private PartitionSelectorStrategy getPartitionSelectorStrategy(ProducerProperties producerProperties) {
+		PartitionSelectorStrategy partitionSelector;
 		if (producerProperties.getPartitionSelectorClass() != null) {
-			return getBean(producerProperties.getPartitionSelectorClass().getName(),
-					PartitionSelectorStrategy.class);
+			logger.warn("'partitionSelectorClass' option is deprecated as of v2.0. Please configure partition "
+					+ "selector as a @Bean that implements 'PartitionSelectorStrategy'. Additionally you can "
+					+ "specify 'spring.cloud.stream.bindings.output.producer.partitionSelectorName' to specify which "
+					+ "bean to use in the event there are more then one.");
+			partitionSelector = instantiate(producerProperties.getPartitionSelectorClass(), PartitionSelectorStrategy.class);
+		}
+		else if (StringUtils.hasText(producerProperties.getPartitionSelectorName())) {
+			partitionSelector = this.partitionSelectors.get(producerProperties.getPartitionSelectorName());
+			Assert.notNull(partitionSelector, "PartitionSelectorStrategy bean with the name '" + producerProperties.getPartitionSelectorName() 
+				+ "' can not be found. Has it been configured (e.g., @Bean)?");
 		}
-		return new DefaultPartitionSelector();
+		else {
+			Assert.isTrue(CollectionUtils.isEmpty(this.partitionSelectors) || this.partitionSelectors.size() == 1, 
+					"Multiple  beans of type 'PartitionSelectorStrategy' found. " + this.partitionSelectors + ". Please "
+							+ "use 'spring.cloud.stream.bindings.output.producer.partitionSelectorName' property to specify "
+							+ "the name of the bean to be used.");		
+			partitionSelector = CollectionUtils.isEmpty(this.partitionSelectors) 
+					? new DefaultPartitionSelector() : this.partitionSelectors.values().iterator().next();
+		}	
+		return partitionSelector;
 	}
 
 	@SuppressWarnings("unchecked")
-	private <T> T getBean(String className, Class<T> type) {
-		if (this.beanFactory.containsBean(className)) {
-			return this.beanFactory.getBean(className, type);
+	private <T> T instantiate(Class<?> implClass, Class<T> type) {
+		try {
+			return (T) implClass.newInstance();
 		}
-		else {
-			synchronized (this) {
-				T bean;
-				Class<?> clazz;
-				try {
-					clazz = ClassUtils.forName(className, this.beanFactory.getBeanClassLoader());
```

Comment:
Why switch to `newInstance()` instead of `ClassUtils`?

Reply:
If you look a the code and how it was before they already had `Class` (see ProducerProperties) which they were passing as `class.getName` and then creating a class again. So all that it really needed is new instance on the already existing class

Reply:
Doh - yes, of course.